src/language/dafnyInstallation.ts

```diff
@@ -1,477 +1,163 @@
-    await mkdirAsync(installationPath.fsPath, { recursive: true });
-    this.writeStatus(`Installing Dafny from source in ${installationPath.fsPath}.\n`);
-    const previousDirectory = processCwd();
-    processChdir(installationPath.fsPath);
-    try {
-      await checkSupportedDotnetVersion();
-    } catch(error: unknown) {
-      try {
-        this.writeStatus('dotnet not found in $PATH, trying to install from brew.');
-        await this.execLog('brew install dotnet-sdk');
-      } catch(error: unknown) {
-        this.writeStatus('An error occurred while running this command.');
-        this.writeStatus(`${error}`);
-        this.writeStatus(`If brew is installed on your system, this can usually be resolved by adding add all brew commands to your ~/.zprofile,
-        e.g. by running the script there https://apple.stackexchange.com/a/430904 :
-  
-        > echo 'eval "$(/opt/homebrew/bin/brew shellenv)"' >> ~/.zprofile
-        > eval "$(/opt/homebrew/bin/brew shellenv)"
-  
-        and restart VSCode, which may reinstall Dafny.`);
-        return false;
-      }
-    }
-    const configuredVersion = await getConfiguredVersion(this.context);
-    if(versionToNumeric(configuredVersion) < versionToNumeric('3.9.0')) {
-      try {
-
-        const process = await this.execLog('javac -version');
-        if(!(/javac \d+\.\d+/.exec(process.stdout))
-          && !(/javac \d+\.\d+/.exec(process.stderr))) {
-          throw '';
-        }
-      } catch(error: unknown) {
-        const errorMsg = error === '' ? 'Javac not found' : `${error}`;
-        this.writeStatus(`${errorMsg}. Javac is needed because you use a version of Dafny older than 3.9.0. Please install a valid JDK`
-        + ' and ensure that the path containing javac is in the PATH environment variable. '
-        + 'You can obtain a free open-source JDK 1.8 from here: '
-        + 'https://aws.amazon.com/corretto/');
-        return false;
-      }
+  private async getDotnetToolVersion(): Promise<string> {
+    if(this.toolVersionCache === undefined) {
+      this.toolVersionCache = await DafnyInstaller.getDotnetToolVersionUncached();
     }
+    return this.toolVersionCache;
+  }
 
-    // Clone the right version
-    await this.execLog(`git clone -b v${configuredVersion} --depth 1 --recurse-submodules ${LanguageServerConstants.DafnyGitUrl}`);
-    processChdir(Utils.joinPath(installationPath, 'dafny').fsPath);
+  private static async getDotnetToolVersionUncached(): Promise<string> {
+    const { path: dotnetExecutable } = await getDotnetExecutablePath();
+    const { stdout } = await execFileAsync(dotnetExecutable, [ 'tool', 'search', 'Dafny', '--detail', '--prerelease' ]);
+    const entries = stdout.split('----------------').map(entry => entry.split('\n').filter(e => e !== ''));
+    const dafnyEntry = entries.filter(entry => entry[0] === 'dafny')[0];
+    const versions = dafnyEntry.slice(8).map(versionLine => versionLine.trimStart().split(' ')[0]);
```

Comment:
Rather than slice(8), could you write something like slice("astring".length)? That way, we would know what you are slicing away.